### PR TITLE
Hotfixes

### LINKED ISSUE #6: Forgetti should ignore specific identifier
Given input:

```jsx
export default function Example() {
  if (process.env.NODE_ENV === 'production') {
    return <div />;
  }

  if (import.meta.env.DEV) {
    return <h1 />;
  }

  return <br />;
}
```

Forgetti produces:

```jsx
import { jsx as _jsx } from "react/jsx-runtime";
import { useMemo as _useMemo } from "react";
import { $$cache as _$$cache } from "forgetti/runtime";
export default function Example() {
    var _c, _ref, _c1, _ref1, _c2, _ref2, _c3, _ref3, _c4, _ref4;
    let _c5 = _$$cache(_useMemo, 7), _v = (_c = _c5)[_ref = 0] || (_c[_ref] = process.env), _v2 = (_c1 = _c5)[_ref1 = 1] || (_c1[_ref1] = _v.NODE_ENV);
    if (_v2 === 'production') {
        var _c6, _ref5, _c21, _ref6;
        let _c2 = (_c6 = _c5)[_ref5 = 2] || (_c6[_ref5] = new Array(1)), _v3 = (_c21 = _c2)[_ref6 = 0] || (_c21[_ref6] = /*#__PURE__*/ _jsx("div", {}));
        return _v3;
    }
    let _v4 = (_c2 = _c5)[_ref2 = 3] || (_c2[_ref2] = import.meta.env), _v5 = (_c3 = _c5)[_ref3 = 4] || (_c3[_ref3] = _v4.DEV);
    if (_v5) {
        var _c7, _ref7, _c31, _ref8;
        let _c3 = (_c7 = _c5)[_ref7 = 5] || (_c7[_ref7] = new Array(1)), _v6 = (_c31 = _c3)[_ref8 = 0] || (_c31[_ref8] = /*#__PURE__*/ _jsx("h1", {}));
        return _v6;
    }
    let _v7 = (_c4 = _c5)[_ref4 = 6] || (_c4[_ref4] = /*#__PURE__*/ _jsx("br", {}));
    return _v7;
}
```

Where `process.env.NODE_ENV` and `import.meta.env.DEV` are processed by forgetti.

Currently, I am using webpack's `DefinePlugin` to replace `process.env.NODE_ENV` and `import.meta.env.DEV`. In webpack, `DefinePlugin` is executed after loaders (`swc-loaders`, `babel-loaders`, `esbuild-loaders`, etc., and in this case, `forgetti-loader`). `process.env.NODE_ENV` and `import.meta.env.DEV` can no longer be replaced by webpack's `DefinePlugin` after the forgetti transformation.

I propose that forgetti should accept an option to ignore specific identifiers when memoizing.

### LINKED ISSUE #7: Incorrect transformation of useCallback
**Input**

```jsx
import { useCallback, useState } from 'react';

export default function Example() {
  const [count, setCount] = useState(0);
  const handleIncrement = useCallback(() => setCount(c => c + 1), []);

  return (
    <button onClick={handleIncrement}>{count}</button>
  );
}
```

**Output**

```jsx
import { jsx as _jsx } from "react/jsx-runtime";
import { useMemo as _useMemo } from "react";
import { $$cache as _$$cache } from "forgetti/runtime";
import { $$equals as _$$equals } from "forgetti/runtime";
import { useCallback } from 'react';
import { useState } from 'react';
export default function Example() {
    var _c, _ref;
    let _c1 = _$$cache(_useMemo, 5);
    const [count, setCount] = useState(0);
    let _v = (_c = _c1)[_ref = 0] || (_c[_ref] = []), _eq = _$$equals(_c1[1], ()=>setCount((c)=>c + 1)), _v2 = _eq ? _c1[1] : _c1[1] = ()=>setCount((c)=>c + 1);
    const handleIncrement = _v2;
    let _eq2 = _$$equals(_c1[2], handleIncrement), _v3 = _eq2 ? _c1[2] : _c1[2] = handleIncrement, _eq3 = _$$equals(_c1[3], count), _v4 = _eq3 ? _c1[3] : _c1[3] = count, _eq4 = _eq2 && _eq3, _v5 = _eq4 ? _c1[4] : _c1[4] = /*#__PURE__*/ _jsx("button", {
        onClick: _v3,
        children: _v4
    });
    return _v5;
}
```

Notice the useCallback transformation:

```js
let _eq = _$$equals(_c1[1], ()=>setCount((c)=>c + 1));
let _v2 = _eq ? _c1[1] : _c1[1] = ()=>setCount((c)=>c + 1);
const handleIncrement = _v2;
```

Whenever React re-renders, `()=>setCount((c)=>c + 1)` is always re-created, thus it will always have a different reference with `_c1[1]`, and result in `_eq` will always be `false`.

### LINKED ISSUE #9: `const App = function App()` break react hook rule
https://codesandbox.io/p/sandbox/serene-dawn-wgjpfy 
click any button will cause react crash

### PR DESCRIPTION
**Pull request by lxsmnsyc**
- Fix types
- Fix #7 
- Fix optimization step for `memo`, `callback` and `effect` if the expression lacks dependency lists
- Fix optimization conflict between `FunctionExpression` and `VariableDeclarator`
- Fix #9 
- Fix #6

### CODE REVIEW

File: packages/forgetti/test/__snapshots__/hooks.test.ts.snap [hunk 1]
```diff
@@ -53,6 +104,33 @@ function Example(props) {
...
 
+exports[`hooks > should optimize useMemo with 0 dependencies 1`] = `
+"import { useMemo as _useMemo } from \\"react\\";
+import { $$cache as _$$cache } from \\"forgetti/runtime\\";
+import { useMemo } from 'react';
+function Example(props) {
+  let _c = _$$cache(_useMemo, 2),
+    _v = _c[0] ||= [],
+    _v2 = _c[1] ||= (() => props.value())();
+  return _v2;
```
Comment (SukkaW) [inline packages/forgetti/test/__snapshots__/hooks.test.ts.snap:115]: IMHO the transformation could be further optimized:

```diff
- let _c = _$$cache(_useMemo, 2),
+ let _c = _$$cache(_useMemo, 1),
-    _v = _c[0] ||= [],
-    _v2 = _c[1] ||= (() => props.value())();
+    _v1 = _c[0] ||= (() => props.value())();
- return _v2;
+ return _v1;
```

File: packages/forgetti/test/__snapshots__/hooks.test.ts.snap [hunk 2]
```diff
@@ -35,6 +61,31 @@ function Example(props) {
...
+exports[`hooks > should optimize useEffect with 0 dependencies 1`] = `
+"import { useMemo as _useMemo } from \\"react\\";
+import { $$cache as _$$cache } from \\"forgetti/runtime\\";
+import { useEffect } from 'react';
+function Example(props) {
+  let _c = _$$cache(_useMemo, 1),
+    _v = _c[0] ||= [];
+  useEffect(() => props.value(), [_v]);
+}"
+`;
```
Comment (SukkaW) [inline packages/forgetti/test/__snapshots__/hooks.test.ts.snap:73]: IMHO, useEffect with 0 dependency doesn't need to be transformed at all.

File: packages/forgetti/test/__snapshots__/hooks.test.ts.snap (hunk 1 above)
Comment (lxsmnsyc) [reply packages/forgetti/test/__snapshots__/hooks.test.ts.snap:115]: There's no "easy" way to know if the array would be empty. I'm just aggressively optimizing the `dependency` assuming that the value can be arbitrary (for example, `useMemo(() => example, myDeps)`

File: packages/forgetti/test/__snapshots__/hooks.test.ts.snap (hunk 2 above)
Comment (lxsmnsyc) [reply packages/forgetti/test/__snapshots__/hooks.test.ts.snap:73]: 0 dependency transformation means auto-dependency transformation. There's no way to automatically track dependencies for `effect` hooks at the moment, and this change practically aligns to `callback` and `memo` hooks

Comment (lxsmnsyc) [reply packages/forgetti/test/__snapshots__/hooks.test.ts.snap:73]: Same reason as the previous comment.

File: packages/forgetti/test/__snapshots__/hooks.test.ts.snap (hunk 1 above)
Comment (SukkaW) [reply packages/forgetti/test/__snapshots__/hooks.test.ts.snap:115]: I see, thanks!

File: packages/forgetti/test/__snapshots__/hooks.test.ts.snap (hunk 2 above)
Comment (SukkaW) [reply packages/forgetti/test/__snapshots__/hooks.test.ts.snap:73]: I see, thanks!

**Pull request merged**